File: 3. Behavioral Design Patterns/Iterator/Java/IteratorTemplate.java
// Iterator Pattern – Java Template

import java.util.*;

interface Iterator<T> { boolean hasNext(); T next(); }
interface Aggregate<T> { Iterator<T> createIterator(); }

class ConcreteIterator<T> implements Iterator<T> {
    private final List<T> items;
    private int index = 0;
    public ConcreteIterator(List<T> items) { this.items = items; }
    public boolean hasNext() { return index < items.size(); }
    public T next() { return items.get(index++); }
}

class ConcreteAggregate<T> implements Aggregate<T> {
    private final List<T> items = new ArrayList<>();
    public void add(T item) { items.add(item); }
    public Iterator<T> createIterator() { return new ConcreteIterator<>(items); }
}

public class IteratorTemplate {
    public static void main(String[] args) {
        ConcreteAggregate<String> aggregate = new ConcreteAggregate<>();
        aggregate.add("Item1");
        aggregate.add("Item2");
        Iterator<String> iterator = aggregate.createIterator();
        while (iterator.hasNext()) { String item = iterator.next(); }
    }
}

File: 3. Behavioral Design Patterns/Iterator/Kotlin/IteratorTemplate.kt
// Iterator Pattern – Kotlin Template

interface Iterator<T> { fun hasNext(): Boolean; fun next(): T }
interface Aggregate<T> { fun createIterator(): Iterator<T> }

class ConcreteIterator<T>(private val items: List<T>) : Iterator<T> {
    private var index = 0
    override fun hasNext() = index < items.size
    override fun next() = items[index++]
}

class ConcreteAggregate<T> : Aggregate<T> {
    private val items = mutableListOf<T>()
    fun add(item: T) = items.add(item)
    override fun createIterator(): Iterator<T> = ConcreteIterator(items)
}

fun main() {
    val aggregate = ConcreteAggregate<String>()
    aggregate.add("Item1")
    aggregate.add("Item2")
    val iterator = aggregate.createIterator()
    while (iterator.hasNext()) { val item = iterator.next() }
}
